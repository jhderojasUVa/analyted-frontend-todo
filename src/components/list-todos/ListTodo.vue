<script>
import TodoItem from '../todo-item/TodoItem.vue';

export default {
    name: 'ListTodo',
    props: {
        todoItems: Array
    },
    watch: {
        todoItems: (newVal, oldVal) => {
            console.log(newVal);
        }
    },
    methods: {
        deleteTodo(id) {
            // pass to parent
            this.$emit('deleteTodo', id);   
        },
        descriptionChanged(data) {
            // pass to parent
            this.$emit('descriptionChanged', data);
        }
    },
    components: { 
        TodoItem 
    },
    emits: [ 'deleteTodo', 'descriptionChanged' ]
}
</script>

<template>
    <ul>
        <li v-for="item in todoItems">
         <TodoItem :data="item" @deleteTodo="deleteTodo" @descriptionChanged="descriptionChanged"/>
        </li>
    </ul>
</template>

<style scoped>
    ul {
        margin: 0;
        padding: 0;
    }

    li {
        list-style-type: none;
        margin: 0.5em 0em;
        padding: 1em;
        border-radius: 3px; 
    }

    li:nth-child(even) {
        background-color: rgba(0, 0, 0, 0.05);
    }

    li a {
        cursor: pointer;
    }
</style>
