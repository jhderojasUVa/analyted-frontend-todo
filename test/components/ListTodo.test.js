import { describe, it, expect } from 'vitest';
import { mount } from '@vue/test-utils';

import ListTodo from '../../src/components/list-todos/ListTodo.vue';

describe('ListTodo', () => {
    it('renders', () => {
        const todoItems = [];

        const wrapper = mount(ListTodo, { props: { todoItems } });

        expect(wrapper).toMatchSnapshot();
    });
    describe('passtrough methods', () => {
        it('deleteTodo', () => {
            const itemID = 1;
            const todoItems = [];

            const wrapper = mount(ListTodo, { props: { todoItems } });

            wrapper.vm.$emit('deleteTodo', itemID);

            expect(wrapper.emitted().deleteTodo).toEqual([[ itemID ]]);
        });
        it('descriptionChanged', () => {
            const data = { id: 1, description: 'foo', completed: false };
            const todoItems = [];

            const wrapper = mount(ListTodo, { props: { todoItems } });

            wrapper.vm.$emit('descriptionChanged', data);

            expect(wrapper.emitted().descriptionChanged).toEqual([[ data ]]);
        });
        it('changeCompleted', () => {
            const data = { id: 1, description: 'foo', completed: false };
            const todoItems = [];

            const wrapper = mount(ListTodo, { props: { todoItems } });

            wrapper.vm.$emit('changeCompleted', data);

            expect(wrapper.emitted().changeCompleted).toEqual([[ data ]]);
        });
    })
});
