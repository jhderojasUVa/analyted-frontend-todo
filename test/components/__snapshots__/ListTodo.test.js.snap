// Vitest Snapshot v1

exports[`ListTodo > renders 1`] = `
VueWrapper {
  "__app": {
    "_component": {
      "name": "VTU_ROOT",
      "render": [Function],
    },
    "_container": <div
      data-v-app=""
    >
      <ul
        data-v-cc3ce813=""
      >
        
        
      </ul>
    </div>,
    "_context": {
      "app": [Circular],
      "components": {
        "transition": {
          "name": "transition",
        },
        "transition-group": {
          "name": "transition-group",
        },
      },
      "config": {
        "compilerOptions": {},
        "errorHandler": undefined,
        "globalProperties": {},
        "isNativeTag": [Function],
        "optionMergeStrategies": {},
        "performance": false,
        "warnHandler": undefined,
      },
      "directives": {},
      "emitsCache": WeakMap {},
      "mixins": [
        {
          "beforeCreate": [Function],
        },
      ],
      "optionsCache": WeakMap {},
      "propsCache": WeakMap {},
      "provides": {},
      "reload": [Function],
    },
    "_instance": {
      "a": null,
      "accessCache": {},
      "appContext": {
        "app": [Circular],
        "components": {
          "transition": {
            "name": "transition",
          },
          "transition-group": {
            "name": "transition-group",
          },
        },
        "config": {
          "compilerOptions": {},
          "errorHandler": undefined,
          "globalProperties": {},
          "isNativeTag": [Function],
          "optionMergeStrategies": {},
          "performance": false,
          "warnHandler": undefined,
        },
        "directives": {},
        "emitsCache": WeakMap {},
        "mixins": [
          {
            "beforeCreate": [Function],
          },
        ],
        "optionsCache": WeakMap {},
        "propsCache": WeakMap {},
        "provides": {},
        "reload": [Function],
      },
      "asyncDep": null,
      "asyncResolved": false,
      "attrs": {},
      "bc": null,
      "bm": null,
      "bu": null,
      "bum": null,
      "c": null,
      "components": null,
      "ctx": {},
      "da": null,
      "data": {},
      "directives": null,
      "ec": null,
      "effect": ReactiveEffect {
        "active": true,
        "allowRecurse": true,
        "deps": [
          Set {
            [Circular],
          },
          Set {
            [Circular],
          },
          Set {
            [Circular],
          },
          Set {
            [Circular],
          },
        ],
        "fn": [Function],
        "onTrack": undefined,
        "onTrigger": undefined,
        "parent": undefined,
        "scheduler": [Function],
      },
      "emit": [Function],
      "emitsOptions": null,
      "emitted": null,
      "exposeProxy": null,
      "exposed": null,
      "inheritAttrs": undefined,
      "isDeactivated": false,
      "isMounted": true,
      "isUnmounted": false,
      "m": null,
      "next": null,
      "parent": null,
      "props": {},
      "propsDefaults": {},
      "propsOptions": [
        {},
        [],
      ],
      "provides": {},
      "proxy": {},
      "refs": {
        "VTU_COMPONENT": {
          "changeCompleted": [Function],
          "deleteTodo": [Function],
          "descriptionChanged": [Function],
          "hasOwnProperty": [Function],
          "todoItems": [],
        },
      },
      "render": [Function],
      "renderCache": [],
      "root": [Circular],
      "rtc": null,
      "rtg": null,
      "scope": EffectScope {
        "active": true,
        "cleanups": [],
        "effects": [
          ReactiveEffect {
            "active": true,
            "allowRecurse": true,
            "deps": [
              Set {
                [Circular],
              },
              Set {
                [Circular],
              },
              Set {
                [Circular],
              },
              Set {
                [Circular],
              },
            ],
            "fn": [Function],
            "onTrack": undefined,
            "onTrigger": undefined,
            "parent": undefined,
            "scheduler": [Function],
          },
        ],
      },
      "setupContext": null,
      "setupState": {},
      "slots": {},
      "sp": null,
      "subTree": {
        "__v_isVNode": true,
        "__v_skip": true,
        "anchor": null,
        "appContext": null,
        "children": null,
        "component": {
          "a": null,
          "accessCache": {
            "@@__IMMUTABLE_ITERABLE__@@": 0,
            "@@__IMMUTABLE_RECORD__@@": 0,
            "_isMockFunction": 0,
            "changeCompleted": 4,
            "constructor": 0,
            "deleteTodo": 4,
            "descriptionChanged": 4,
            "hasOwnProperty": 4,
            "toJSON": 0,
            "todoItems": 3,
            Symbol(Symbol.toStringTag): 0,
          },
          "appContext": {
            "app": [Circular],
            "components": {
              "transition": {
                "name": "transition",
              },
              "transition-group": {
                "name": "transition-group",
              },
            },
            "config": {
              "compilerOptions": {},
              "errorHandler": undefined,
              "globalProperties": {},
              "isNativeTag": [Function],
              "optionMergeStrategies": {},
              "performance": false,
              "warnHandler": undefined,
            },
            "directives": {},
            "emitsCache": WeakMap {},
            "mixins": [
              {
                "beforeCreate": [Function],
              },
            ],
            "optionsCache": WeakMap {},
            "propsCache": WeakMap {},
            "provides": {},
            "reload": [Function],
          },
          "asyncDep": null,
          "asyncResolved": false,
          "attrs": {},
          "bc": null,
          "bm": null,
          "bu": null,
          "bum": null,
          "c": null,
          "components": {
            "TodoItem": {
              "__file": "/mnt/c/Projects/analyted/analyted-frontend-todo/src/components/todo-item/TodoItem.vue",
              "__scopeId": "data-v-46d2f85e",
              "components": {
                "InlineEditing": {
                  "__file": "/mnt/c/Projects/analyted/analyted-frontend-todo/src/components/inline-editing/InlineEditing.vue",
                  "__scopeId": "data-v-b406172e",
                  "methods": {
                    "updateText": [Function],
                  },
                  "mounted": [Function],
                  "name": "InlineEditing",
                  "props": {
                    "data": [Function],
                  },
                  "render": [Function],
                },
              },
              "data": [Function],
              "emits": [
                "deleteTodo",
                "descriptionChanged",
                "changeCompleted",
              ],
              "methods": {
                "changeComplete": [Function],
                "deleteTodo": [Function],
                "descriptionChanged": [Function],
                "editItem": [Function],
              },
              "name": "TodoItem",
              "props": {
                "data": [Function],
              },
              "render": [Function],
            },
          },
          "ctx": {
            "changeCompleted": [Function],
            "deleteTodo": [Function],
            "descriptionChanged": [Function],
            "hasOwnProperty": [Function],
            "todoItems": [],
          },
          "da": null,
          "data": {},
          "directives": null,
          "ec": null,
          "effect": ReactiveEffect {
            "active": true,
            "allowRecurse": true,
            "deps": [
              Set {
                [Circular],
              },
              Set {
                [Circular],
              },
            ],
            "fn": [Function],
            "onTrack": undefined,
            "onTrigger": undefined,
            "parent": undefined,
            "scheduler": [Function],
          },
          "emit": [Function],
          "emitsOptions": {
            "changeCompleted": null,
            "deleteTodo": null,
            "descriptionChanged": null,
          },
          "emitted": null,
          "exposeProxy": null,
          "exposed": null,
          "inheritAttrs": undefined,
          "isDeactivated": false,
          "isMounted": true,
          "isUnmounted": false,
          "m": null,
          "next": null,
          "parent": [Circular],
          "props": {
            "todoItems": [],
          },
          "propsDefaults": {},
          "propsOptions": [
            {
              "todoItems": {
                "0": false,
                "1": true,
                "type": [Function],
              },
            },
            [],
          ],
          "provides": {},
          "proxy": {
            "changeCompleted": [Function],
            "deleteTodo": [Function],
            "descriptionChanged": [Function],
            "hasOwnProperty": [Function],
            "todoItems": [],
          },
          "refs": {},
          "render": [Function],
          "renderCache": [],
          "root": [Circular],
          "rtc": null,
          "rtg": null,
          "scope": EffectScope {
            "active": true,
            "cleanups": [],
            "effects": [
              ReactiveEffect {
                "active": true,
                "allowRecurse": true,
                "deps": [
                  Set {
                    [Circular],
                  },
                  Set {
                    [Circular],
                  },
                ],
                "fn": [Function],
                "onTrack": undefined,
                "onTrigger": undefined,
                "parent": undefined,
                "scheduler": [Function],
              },
            ],
          },
          "setupContext": null,
          "setupState": {},
          "slots": {},
          "sp": null,
          "subTree": {
            "__v_isVNode": true,
            "__v_skip": true,
            "anchor": null,
            "appContext": null,
            "children": [
              {
                "__v_isVNode": true,
                "__v_skip": true,
                "anchor": ,
                "appContext": null,
                "children": [],
                "component": null,
                "dirs": null,
                "dynamicChildren": [],
                "dynamicProps": null,
                "el": ,
                "key": null,
                "patchFlag": 256,
                "props": null,
                "ref": null,
                "scopeId": "data-v-cc3ce813",
                "shapeFlag": 16,
                "slotScopeIds": null,
                "ssContent": null,
                "ssFallback": null,
                "staticCount": 0,
                "suspense": null,
                "target": null,
                "targetAnchor": null,
                "transition": null,
                "type": Symbol(Fragment),
              },
            ],
            "component": null,
            "dirs": null,
            "dynamicChildren": [
              {
                "__v_isVNode": true,
                "__v_skip": true,
                "anchor": ,
                "appContext": null,
                "children": [],
                "component": null,
                "dirs": null,
                "dynamicChildren": [],
                "dynamicProps": null,
                "el": ,
                "key": null,
                "patchFlag": 256,
                "props": null,
                "ref": null,
                "scopeId": "data-v-cc3ce813",
                "shapeFlag": 16,
                "slotScopeIds": null,
                "ssContent": null,
                "ssFallback": null,
                "staticCount": 0,
                "suspense": null,
                "target": null,
                "targetAnchor": null,
                "transition": null,
                "type": Symbol(Fragment),
              },
            ],
            "dynamicProps": null,
            "el": <ul
              data-v-cc3ce813=""
            >
              
              
            </ul>,
            "key": null,
            "patchFlag": 0,
            "props": null,
            "ref": null,
            "scopeId": "data-v-cc3ce813",
            "shapeFlag": 17,
            "slotScopeIds": null,
            "ssContent": null,
            "ssFallback": null,
            "staticCount": 0,
            "suspense": null,
            "target": null,
            "targetAnchor": null,
            "transition": null,
            "type": "ul",
          },
          "suspense": null,
          "suspenseId": 0,
          "type": {
            "__file": "/mnt/c/Projects/analyted/analyted-frontend-todo/src/components/list-todos/ListTodo.vue",
            "__scopeId": "data-v-cc3ce813",
            "components": {
              "TodoItem": {
                "__file": "/mnt/c/Projects/analyted/analyted-frontend-todo/src/components/todo-item/TodoItem.vue",
                "__scopeId": "data-v-46d2f85e",
                "components": {
                  "InlineEditing": {
                    "__file": "/mnt/c/Projects/analyted/analyted-frontend-todo/src/components/inline-editing/InlineEditing.vue",
                    "__scopeId": "data-v-b406172e",
                    "methods": {
                      "updateText": [Function],
                    },
                    "mounted": [Function],
                    "name": "InlineEditing",
                    "props": {
                      "data": [Function],
                    },
                    "render": [Function],
                  },
                },
                "data": [Function],
                "emits": [
                  "deleteTodo",
                  "descriptionChanged",
                  "changeCompleted",
                ],
                "methods": {
                  "changeComplete": [Function],
                  "deleteTodo": [Function],
                  "descriptionChanged": [Function],
                  "editItem": [Function],
                },
                "name": "TodoItem",
                "props": {
                  "data": [Function],
                },
                "render": [Function],
              },
            },
            "emits": [
              "deleteTodo",
              "descriptionChanged",
              "changeCompleted",
            ],
            "methods": {
              "changeCompleted": [Function],
              "deleteTodo": [Function],
              "descriptionChanged": [Function],
            },
            "name": "ListTodo",
            "props": {
              "todoItems": [Function],
            },
            "render": [Function],
          },
          "u": null,
          "uid": 1,
          "um": null,
          "update": [Function],
          "vnode": [Circular],
          "withProxy": null,
        },
        "dirs": null,
        "dynamicChildren": null,
        "dynamicProps": null,
        "el": <ul
          data-v-cc3ce813=""
        >
          
          
        </ul>,
        "key": null,
        "patchFlag": 0,
        "props": {
          "ref": "VTU_COMPONENT",
          "todoItems": [],
        },
        "ref": {
          "f": false,
          "i": [Circular],
          "k": undefined,
          "r": "VTU_COMPONENT",
        },
        "scopeId": null,
        "shapeFlag": 4,
        "slotScopeIds": null,
        "ssContent": null,
        "ssFallback": null,
        "staticCount": 0,
        "suspense": null,
        "target": null,
        "targetAnchor": null,
        "transition": null,
        "type": {
          "__file": "/mnt/c/Projects/analyted/analyted-frontend-todo/src/components/list-todos/ListTodo.vue",
          "__scopeId": "data-v-cc3ce813",
          "components": {
            "TodoItem": {
              "__file": "/mnt/c/Projects/analyted/analyted-frontend-todo/src/components/todo-item/TodoItem.vue",
              "__scopeId": "data-v-46d2f85e",
              "components": {
                "InlineEditing": {
                  "__file": "/mnt/c/Projects/analyted/analyted-frontend-todo/src/components/inline-editing/InlineEditing.vue",
                  "__scopeId": "data-v-b406172e",
                  "methods": {
                    "updateText": [Function],
                  },
                  "mounted": [Function],
                  "name": "InlineEditing",
                  "props": {
                    "data": [Function],
                  },
                  "render": [Function],
                },
              },
              "data": [Function],
              "emits": [
                "deleteTodo",
                "descriptionChanged",
                "changeCompleted",
              ],
              "methods": {
                "changeComplete": [Function],
                "deleteTodo": [Function],
                "descriptionChanged": [Function],
                "editItem": [Function],
              },
              "name": "TodoItem",
              "props": {
                "data": [Function],
              },
              "render": [Function],
            },
          },
          "emits": [
            "deleteTodo",
            "descriptionChanged",
            "changeCompleted",
          ],
          "methods": {
            "changeCompleted": [Function],
            "deleteTodo": [Function],
            "descriptionChanged": [Function],
          },
          "name": "ListTodo",
          "props": {
            "todoItems": [Function],
          },
          "render": [Function],
        },
      },
      "suspense": null,
      "suspenseId": 0,
      "type": {
        "name": "VTU_ROOT",
        "render": [Function],
      },
      "u": null,
      "uid": 0,
      "um": null,
      "update": [Function],
      "vnode": {
        "__v_isVNode": true,
        "__v_skip": true,
        "anchor": null,
        "appContext": {
          "app": [Circular],
          "components": {
            "transition": {
              "name": "transition",
            },
            "transition-group": {
              "name": "transition-group",
            },
          },
          "config": {
            "compilerOptions": {},
            "errorHandler": undefined,
            "globalProperties": {},
            "isNativeTag": [Function],
            "optionMergeStrategies": {},
            "performance": false,
            "warnHandler": undefined,
          },
          "directives": {},
          "emitsCache": WeakMap {},
          "mixins": [
            {
              "beforeCreate": [Function],
            },
          ],
          "optionsCache": WeakMap {},
          "propsCache": WeakMap {},
          "provides": {},
          "reload": [Function],
        },
        "children": null,
        "component": [Circular],
        "dirs": null,
        "dynamicChildren": null,
        "dynamicProps": null,
        "el": <ul
          data-v-cc3ce813=""
        >
          
          
        </ul>,
        "key": null,
        "patchFlag": 0,
        "props": null,
        "ref": null,
        "scopeId": null,
        "shapeFlag": 4,
        "slotScopeIds": null,
        "ssContent": null,
        "ssFallback": null,
        "staticCount": 0,
        "suspense": null,
        "target": null,
        "targetAnchor": null,
        "transition": null,
        "type": {
          "name": "VTU_ROOT",
          "render": [Function],
        },
      },
      "withProxy": null,
    },
    "_props": null,
    "_uid": 0,
    "component": [Function],
    "config": {
      "compilerOptions": {},
      "errorHandler": undefined,
      "globalProperties": {},
      "isNativeTag": [Function],
      "optionMergeStrategies": {},
      "performance": false,
      "warnHandler": undefined,
    },
    "directive": [Function],
    "mixin": [Function],
    "mount": [Function],
    "provide": [Function],
    "unmount": [Function],
    "use": [Function],
    "version": "3.2.37",
  },
  "__setProps": [Function],
  "componentVM": {
    "changeCompleted": [Function],
    "deleteTodo": [Function],
    "descriptionChanged": [Function],
    "hasOwnProperty": [Function],
    "todoItems": [],
  },
  "isDisabled": [Function],
  "rootVM": {},
  "wrapperElement": <ul
    data-v-cc3ce813=""
  >
    
    
  </ul>,
}
`;
